src/core/DY.Core.Trace.SplitUtils.fst

```diff
@@ -0,0 +1,184 @@
+/// exist, but it is easy to generalize to find the oldest such entry, or all such entries.
+///
+/// Using trace entry equivalence and this first search function, we can find entries that are
+/// equivalent to a known trace entry, though they may not necessarily be equal.
+
+[@@"opaque_to_smt"]
+let rec trace_search (#label_t:Type) (tr:trace_ label_t) (p:trace_entry_ label_t -> bool)
+  : Pure (option timestamp)
+    (requires True)
+    (ensures fun ts_opt ->
+       match ts_opt with
+       | None -> forall ts. ts `on_trace` tr ==> ~(p(get_entry_at tr ts))
+       | Some ts -> ts `on_trace` tr /\ p (get_entry_at tr ts)
+    )
+  = match tr with
+    | Nil -> None
+    | Snoc hd entry ->
+      if p entry then
+        Some (last_ts tr)
+      else
+        trace_search hd p
+
+[@@"opaque_to_smt"]
+let trace_find (#label_t:Type) (tr:trace_ label_t) (e:trace_entry_ label_t{entry_exists tr e})
+  : Pure timestamp (requires True) (ensures fun ts -> ts `on_trace` tr /\ trace_entry_equiv e (get_entry_at tr ts))
+  = let Some ts = trace_search tr (trace_entry_equiv e) in
+    ts
+
+(*** Trace arithmetic ***)
+
+/// We define trace concatenation (the monoid operation), as well as
+/// trace subtraction, which is defined for traces tr2 <$ tr1, and yields the suffix of
+/// tr1 beginning after tr2.
+
+[@@"opaque_to_smt"]
+let rec trace_concat (#label_t:Type) (tr1 tr2:trace_ label_t)
+  : trace_ label_t
+  = match tr2 with
+    | Nil -> tr1
+    | Snoc hd e -> append_entry (trace_concat tr1 hd) e
+
+[@@"opaque_to_smt"]
+let rec trace_subtract (#label_t:Type) (tr1:trace_ label_t) (tr2:trace_ label_t{tr2 <$ tr1})
+  : trace_ label_t
+  = if trace_length tr1 = trace_length tr2 then Nil
+    else begin
+      let Snoc hd e = tr1 in
+      grows_cases tr2 tr1;
+      append_entry (trace_subtract hd tr2) e
+    end
+
+let (<++>) = trace_concat
+let (<-->) = trace_subtract
+
+/// Traces form a monoid with trace concatenation as operation, and Nil as unit.
+
+let rec trace_concat_nil (#label_t:Type) (tr:trace_ label_t)
+  : Lemma
+    (
+      tr <++> empty_trace == tr /\
+      empty_trace <++> tr == tr
+    )
+    [SMTPatOr [[SMTPat (tr <++> empty_trace)]; [SMTPat (empty_trace <++> tr)]]]
+  = norm_spec [zeta; delta_only [`%trace_concat]] (trace_concat #label_t);
+    match tr with
+    | Nil -> ()
+    | Snoc hd _ -> trace_concat_nil hd
+
+let rec trace_concat_assoc (#label_t:Type) (tr1 tr2 tr3:trace_ label_t)
+  : Lemma
+    (((tr1 <++> tr2) <++> tr3) == (tr1 <++> (tr2 <++> tr3)))
+  = norm_spec [zeta; delta_only [`%trace_concat]] (trace_concat #label_t);
+    match tr3 with
+    | Nil -> ()
+    | Snoc hd _ -> trace_concat_assoc tr1 tr2 hd
+
+let snoc_is_concat_singleton (#label_t:Type) (tr :trace_ label_t) (e:trace_entry_ label_t)
+  : Lemma
+    (append_entry tr e == (tr <++> (Snoc Nil e)))
+    [SMTPat (tr <++> (Snoc Nil e))]
+  = norm_spec [zeta; delta_only [`%trace_concat]] (trace_concat #label_t)
```

Comment:
nit: why not `append_entry empty_trace e`?

Reply:
Here, I had tried with `trace_from_list [e]`, which seemed like the most natural way to show a singleton, and couldn't get it to work nicely. `append_entry empty_trace e` seems to work, but is of course more verbose.

Reply:
I think if you write `trace_from_list [e]` then you need a fuel/ifuel of 1 (or maybe 2). If possible I would rather avoid writing `Snoc` and `Nil` when possible now that we have `empty_trace` and `append_entry`.